lib_eio_posix/process.ml

```diff
@@ -0,0 +1,72 @@
+open Low_level
+module Rfcd = Eio_unix.Private.Rcfd
+
+let unix_status_to_stats = function
+  | Unix.WEXITED i -> Eio.Process.Exited i
+  | Unix.WSIGNALED i -> Eio.Process.Signaled i
+  | Unix.WSTOPPED i -> Eio.Process.Stopped i
+
+let process proc : Eio.Process.t = object
+  method pid = Process.pid proc
+  method await =
+    let status = Eio.Promise.await @@ Process.exit_status proc in
+    unix_status_to_stats status
+  method signal i = Process.signal proc i
+end
+
+let read_of_fd ~sw flow =
+  match Fd.get_fd_opt flow with
+  | None ->
+    let r, w = pipe ~sw in
+    Some (Flow.of_fd w), r
+  | Some fd -> None, fd
+
+let write_of_fd ~sw t =
+  match Fd.get_fd_opt t with
+  | None ->
+    let r, w = pipe ~sw in
+    Some (Flow.of_fd r), w
+  | Some fd -> None, fd
+
+let v = object
+  inherit Eio.Process.mgr
+
+  method spawn ~sw ?cwd ~(stdin : #Eio.Flow.source) ~(stdout : #Eio.Flow.sink) ~(stderr : #Eio.Flow.sink) prog args =
+    let stdin_w, stdin_fd = read_of_fd ~sw stdin in
+    let stdout_r, stdout_fd = write_of_fd ~sw stdout in
+    let stderr_r, stderr_fd = write_of_fd ~sw stderr in
+    let actions = Process.Fork_action.[
+      Eio_unix.Private.Fork_action.inherit_fds [
+        0, Fd.to_rcfd stdin_fd, `Blocking;
+        1, Fd.to_rcfd stdout_fd, `Blocking;
+        2, Fd.to_rcfd stderr_fd, `Blocking;
+      ];
+      execve prog ~argv:(Array.of_list args) ~env:[||]
+    ] in
+    let with_actions cwd fn = match cwd with
+      | Some ((dir, path) : Eio.Fs.dir Eio.Path.t) -> (
+        match Eio.Generic.probe dir Fs.Posix_dir with
+        | None -> fn actions
+        | Some posix ->
+          posix#with_parent_dir path @@ fun dirfd s ->
+          let cwd = Low_level.openat ?dirfd ~sw ~mode:0 s Low_level.Open_flags.(rdonly + directory) in
+          fn (Process.Fork_action.fchdir cwd :: actions)
+      )
+      | None -> fn actions
+    in
+    let proc =
+      with_actions cwd @@ fun actions ->
+      process (Process.spawn ~sw actions)
+    in
+    Option.iter (fun stdin_w ->
+      Eio.Flow.copy stdin stdin_w;
```

Comment:
You need to fork off a another fiber to handle this (I see you had it that way earlier and then removed it for some reason).

Failing to fork will work OK for short messages, where the kernel will buffer them, but will deadlock for longer ones. It's also a problem for interactive programs (e.g. hooked up to GUI). Imagine the program wants to output a long set of terms and conditions and then ask for confirmation. Once the stdout pipe is full, it will stop and wait for us to read it, but we're still here waiting for the user to finish sending the input (which they won't do because they haven't seen any output yet).

Reply:
Very quickly, it seemed that there was a race between the process exiting and the standard input/output being read/written. This very reliably fails in the old fork version: 

```ocaml
open Eio
open Eio.Std

let run env =
  let buf = Buffer.create 128 in
  Switch.run @@ fun sw ->
  let stdout = Flow.buffer_sink buf in
  let child =
    Process.spawn ~sw env#process_mgr ~stdin:env#stdin ~stdout ~stderr:env#stderr "/bin/echo" [ "echo"; "Hello" ]
  in
  Process.await_exn child;
  let contents = Buffer.contents buf in
  if contents = "Hello\n" then () else failwith contents

let () =
  Eio_main.run @@ fun env ->
  for i = 0 to 1000 do
    Eio.traceln "Process number %i" i;
    run env
  done
```

But your comment is right!

Reply:
Yes, you need to wait for stdout to be copied to the buffer. Just waiting for the process to exit isn't sufficient. This works (but is easy to get wrong):
```ocaml
let run env =
  let buf = Buffer.create 128 in
  Switch.run (fun sw ->
      let stdout = Flow.buffer_sink buf in
      let child =
        Process.spawn ~sw env#process_mgr ~stdin:env#stdin ~stdout ~stderr:env#stderr "/bin/echo" [ "echo"; "Hello" ]
      in
      Process.await_exn child;
    );
  let contents = Buffer.contents buf in
  if contents = "Hello\n" then () else failwith contents
```
Perhaps we should just expose the pipes and let the user do the reading, as Lwt does? It's a bit of a pain with the types (which is why Lwt needs so many functions). Maybe GADTs would help (so if you pass `~stdout:Pipe` then you can do `Process.stdout child` to get the flow).

Then we can add convenience functions on top (like `popen`) that collect the output for you, rather than hard-coding that behaviour in each backend.

https://docs.python.org/3/library/subprocess.html might be a good source of ideas.